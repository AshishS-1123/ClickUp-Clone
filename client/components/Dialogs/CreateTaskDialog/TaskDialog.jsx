import React, { useState, useRef } from "react";
import Dialog from "@mui/material/Dialog";
import CloseIcon from "@mui/icons-material/Close";
import PriorityButton from "./PropertiesButtons/PriorityButton";
import TagsButton from "./PropertiesButtons/TagsButton";
import DueDatesButton from "./PropertiesButtons/DueDatesButton";
import StatusButton from "./PropertiesButtons/StatusButton";
import styles from "./TaskDialog.module.css";
import themeColors from "../../../utils/contexts/themeContext";

const iconStyles = {
  color: "#8a8d91",
  width: "34px",
  height: "34px",
  border: "1px dashed #d5d6d7",
  borderRadius: "50%",
  padding: "5px",
  margin: "0 5px",
}

const elementStyles = {
  color: themeColors.textColor,
  background: themeColors.background
}

function CreateTaskDialog({ open, closeDialog, handleCreateTask, inList, forWorkspace }) {
  const taskNameRef = useRef(null);
  const [priority, setPriority] = useState("");
  const [dueDate, setDueDate] = useState(null);
  const [status, setStatus] = useState("");

  const handleSubmitButtonClick = () => {
    handleCreateTask(taskNameRef.current.value, priority, dueDate, status);
  }

  return (
    <Dialog
      open={open}
      onBackdropClick={closeDialog}
      sx={{
        "& .MuiDialog-paper": {
          width: "580px",
          height: "510px",
          position: "absolute",
          bottom: "0px",
          right: "0px",
          background: themeColors.background,
          display: "flex",
          flexDirection: "column",
          padding: "20px",
        },
      }}
    >
      <div className={styles.topBar}>
        <input placeholder='Task name' style={elementStyles} ref={taskNameRef} />
        <CloseIcon
          sx={{ fill: themeColors.textBoldColor, width: "32px", height: "32px" }}
          onClick={closeDialog}
        />
      </div>

      <div className={styles.parentInfoBar}>
        <label>In</label>
        <div className={styles.forList} style={elementStyles}>{inList}</div>

        <label>For</label>
        <div className={styles.forWorkspace} style={elementStyles}>{forWorkspace}</div>
      </div>

      <textarea className={styles.descriptionInput} style={{ ...elementStyles, resize: "none" }} />

      <div className={styles.bottomBar}>
        <div className={styles.propertiesBar}>
          <DueDatesButton onDateSelect={setDueDate} />
          <PriorityButton onPrioritySelect={setPriority} />
          <TagsButton />
          <StatusButton onStatusSelect={setStatus} />
        </div>
        <button
          className={styles.button}
          style={{ background: themeColors.accentColor, color: themeColors.background }}
          onClick={handleSubmitButtonClick}
        >
          Create Task
        </button>
      </div>

    </Dialog>
  );
}

export default CreateTaskDialog;
