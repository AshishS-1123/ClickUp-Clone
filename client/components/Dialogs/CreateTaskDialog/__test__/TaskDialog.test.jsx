import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import TaskDialog from "../TaskDialog";

let component = null;

describe("<TaskDialog /> tests", () => {
  // Snapshot Testing
  it("should match snapshot", () => {
    component = render(
      <TaskDialog
        open
        closeDialog={() => { }}
        handleCreateTask={() => { }}
        ref={null}
      />,
    );

    expect(component.container).toMatchSnapshot();
  });

  // Test for syntax errors.
  it("should render without crashing", () => {
    component = render(
      <TaskDialog
        open
        closeDialog={() => { }}
        handleCreateTask={() => { }}
        ref={null}
      />,
    );
  });

  //Test that visibility of component can be controlled by open prop.
  it("should be displayed when 'open' prop is passed", () => {
    component = render(
      <TaskDialog
        open
        closeDialog={() => { }}
        handleCreateTask={() => { }}
        ref={null}
      />,
    );

    expect(screen.getByRole("dialog")).toBeDefined();
  });

  //Test that visibility of component can be controlled by open prop.
  it("should not be displayed when 'open' prop is passed", () => {
    component = render(
      <TaskDialog
        open={false}
        closeDialog={() => { }}
        handleCreateTask={() => { }}
        ref={null}
      />,
    );

    expect(screen.queryByRole("dialog")).toBeNull();
  });

  // Standard behaviour for backdrop. Implementation from mui.
  it("should close when backdrop is clicked", () => {
    const closeDialog = jest.fn();
    const handleCreateTask = jest.fn();

    component = render(
      <TaskDialog
        open
        closeDialog={closeDialog}
        handleCreateTask={handleCreateTask}
        ref={null}
      />,
    );

    // Get the backdrop.
    const backdrop = screen.getByRole("dialog").parentElement;

    // Click on the backdrop.
    userEvent.click(backdrop);

    // Assert that the closeDialog fundtion was called.
    expect(closeDialog).toHaveBeenCalledTimes(1);
  });

  // Test that function for closing dialog is called when close icon is clicked.
  it("should should close when close icon is clicked", () => {
    const closeDialog = jest.fn();
    const handleCreateTask = jest.fn();

    component = render(
      <TaskDialog
        open
        closeDialog={closeDialog}
        handleCreateTask={handleCreateTask}
        ref={null}
      />,
    );

    // Get the close icon button.
    const button = screen.getByTestId("CloseIcon");

    // Click on the button.
    userEvent.click(button);

    // Assert that the closeDialog fundtion was called.
    expect(closeDialog).toHaveBeenCalledTimes(1);
  });

  // Check UI layout.
  it("should have a title, close icon, input element and submit button", () => {
    component = render(
      <TaskDialog
        open
        closeDialog={jest.fn()}
        handleCreateTask={jest.fn()}
        ref={null}
      />,
    );

    // Check label is present.
    expect(screen.getByLabelText("Task name")).toBeDefined();
    // Check input is present.
    expect(screen.getByPlaceholderText("Enter task name")).toBeDefined();
    // Check submit button is present
    expect(screen.getByText("Create task")).toBeDefined();
  });
});
