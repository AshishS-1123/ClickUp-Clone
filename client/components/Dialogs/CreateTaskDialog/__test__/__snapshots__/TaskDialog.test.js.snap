// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskDialog /> tests should match snapshot 1`] = `
<div
  aria-hidden="true"
/>
`;
